caluma/form/historical_schema.py

```diff
@@ -24,18 +24,25 @@
 def historical_qs_as_of(queryset, date, pk_attr):
     """Get history revision as of `date` for queryset.
 
-    :param queryset: history qs
+    :param queryset: history Queryset()
     :param date: aware datetime()
     :param pk_attr: str (pk field name)
+    :return: Queryset()
     """
-    # TODO: This could be optimised with some sql magic in order to return a queryset.
+    # TODO: This could be optimised with some sql magic.
     # This could become unnecessary as soon as
     # https://github.com/treyhunner/django-simple-history/issues/397 is resolved.
     queryset = queryset.filter(history_date__lte=date)
     for original_pk in set(queryset.values_list(pk_attr, flat=True)):
-        changes = queryset.filter(**{pk_attr: original_pk})
-        last_change = changes.latest("history_date")
-        yield last_change
+        # exclude all records referencing the same original_pk except the most recent
+        queryset = queryset.exclude(
```

Comment:
I don't like this. It creates a whole bunch of nested queries, inverting each other. For each `original_pk`, you will have an additional nesting level, as you use the (updated) queryset variable to filter for the next `pk`.

I think we'd be better off by firing off an intermediate query.. First, get the newest versions using an aggregate (`Max('history_date')`), group by `pk_attr`. The resulting pairs can then be used to build a query using `Q` objects.